File: two-sum/two-sum.rb
=begin
def two_sum(arr, target)

  arr.each_with_index do |num1, idx1|
    arr.each_with_index do |num2, idx2|
      sum = num1 + num2
      return [idx1, idx2] if sum == target && idx1 < idx2
    end
  end
end

=end

def two_sum(arr, target)
  pairs = {}

  arr.each_with_index do |num, idx|
    p pairs
    if pairs[target - num]
      return pairs[target - num], idx
    end
    pairs[num] = idx
  end
end


p two_sum([2,7,11,15], 9) # [0,1]
p two_sum([3,2,4], 6) # [1,2]

File: two_sum/two-sum.js
const twoSum = (arr, target) => {
  for(i = 0; i < arr.length; i++){
    for(j = 0; j < arr.length; j++){
      let sum = arr[i] + arr[j]
      if (sum === target){
        return [i, j]
      }
    }
  }
}


const twoSum = (arr, target) => {
  let sums = {}

  for(i = 0; i < arr.length; i++){
    if(target - arr[i] in sums){
      return [sums[target - arr[i]], i]
    }else {
      sums[arr[i]] = i
    }
  }
 }


console.log(twoSum([2,7,11,15], 9)); // [0, 1]
console.log(twoSum([3,2,4], 6)); // [1,2]
console.log(twoSum([3,3], 6)); // [0, 1]

